### PR TITLE
ref(*): kubernetes v1.9 support

### PR DESCRIPTION
**Pull request by adamreese**

### CODE REVIEW

File: cmd/helm/installer/install.go [hunk 1]
```diff
@@ -30,21 +30,22 @@ import (
...
+
 	"k8s.io/helm/pkg/chartutil"
 )
 
 // Install uses Kubernetes client to install Tiller.
 //
 // Returns an error if the command failed.
 func Install(client kubernetes.Interface, opts *Options) error {
-	if err := createDeployment(client.Extensions(), opts); err != nil {
+	if err := createDeployment(client.ExtensionsV1beta1(), opts); err != nil {
```
Comment (bacongobbler) [inline cmd/helm/installer/install.go]: It was announced at kubecon that the workloads API (deployments, replicasets, statefulsets, daemonsets, etc) will move to apps/v1 in 1.9, so this probably needs to be updated.

https://github.com/kubernetes/kubernetes/tree/v1.9.0-beta.2/staging/src/k8s.io/client-go/kubernetes/typed/apps/v1

File: cmd/helm/installer/install.go [hunk 2]
```diff
@@ -55,21 +56,21 @@ func Install(client kubernetes.Interface, opts *Options) error {
 //
 // Returns an error if the command failed.
 func Upgrade(client kubernetes.Interface, opts *Options) error {
-	obj, err := client.Extensions().Deployments(opts.Namespace).Get(deploymentName, metav1.GetOptions{})
+	obj, err := client.ExtensionsV1beta1().Deployments(opts.Namespace).Get(deploymentName, metav1.GetOptions{})
```
Comment (bacongobbler) [inline cmd/helm/installer/install.go]: same here RE: kubecon

File: cmd/helm/installer/install.go [hunk 3]
```diff
@@ -55,21 +56,21 @@ func Install(client kubernetes.Interface, opts *Options) error {
...
 	if err != nil {
 		return err
 	}
 	obj.Spec.Template.Spec.Containers[0].Image = opts.selectImage()
 	obj.Spec.Template.Spec.Containers[0].ImagePullPolicy = opts.pullPolicy()
 	obj.Spec.Template.Spec.ServiceAccountName = opts.ServiceAccount
-	if _, err := client.Extensions().Deployments(opts.Namespace).Update(obj); err != nil {
+	if _, err := client.ExtensionsV1beta1().Deployments(opts.Namespace).Update(obj); err != nil {
```
Comment (bacongobbler) [inline cmd/helm/installer/install.go]: same here RE: kubecon

File: pkg/kube/client_test.go [hunk 4]
```diff
@@ -232,50 +232,35 @@ func TestUpdate(t *testing.T) {
...
 		// Test for an invalid manifest
 		infos, err := c.Build(tt.namespace, tt.reader)
-		if err != nil && err.Error() != tt.errMessage {
-			t.Errorf("%q. expected error message: %v, got %v", tt.name, tt.errMessage, err)
-		} else if err != nil && !tt.err {
-			t.Errorf("%q. Got error message when no error should have occurred: %v, got %v", tt.name, tt.errMessage, err)
+		if err != nil && !tt.err {
+			t.Errorf("%q. Got error message when no error should have occurred: %v", tt.name, err)
+		} else if err != nil && strings.Contains(err.Error(), "--validate=false") {
+			t.Errorf("%q. error message was not scrubed", tt.name)
```
Comment (bacongobbler) [inline pkg/kube/client_test.go]: typo: scrubbed

File: pkg/kube/wait.go [hunk 5]
```diff
@@ -67,13 +67,13 @@ func (c *Client) waitForResources(timeout time.Duration, created Result) error {
...
 			case *v1.Pod:
-				pod, err := kcs.Core().Pods(value.Namespace).Get(value.Name, metav1.GetOptions{})
+				pod, err := kcs.CoreV1().Pods(value.Namespace).Get(value.Name, metav1.GetOptions{})
 				if err != nil {
 					return false, err
 				}
 				pods = append(pods, *pod)
 			case *extensions.Deployment:
-				currentDeployment, err := kcs.Extensions().Deployments(value.Namespace).Get(value.Name, metav1.GetOptions{})
+				currentDeployment, err := kcs.ExtensionsV1beta1().Deployments(value.Namespace).Get(value.Name, metav1.GetOptions{})
```
Comment (bacongobbler) [inline pkg/kube/wait.go]: same here RE: kubecon

File: pkg/kube/wait.go [hunk 6]
```diff
@@ -67,13 +67,13 @@ func (c *Client) waitForResources(timeout time.Duration, created Result) error {
 				}
 				pods = append(pods, list...)
 			case *v1.Pod:
-				pod, err := kcs.Core().Pods(value.Namespace).Get(value.Name, metav1.GetOptions{})
+				pod, err := kcs.CoreV1().Pods(value.Namespace).Get(value.Name, metav1.GetOptions{})
 				if err != nil {
 					return false, err
 				}
 				pods = append(pods, *pod)
 			case *extensions.Deployment:
```
Comment (bacongobbler) [inline pkg/kube/wait.go]: we'll probably need to add the appsv1.* types in the watcher here as well.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: cmd/helm/installer/install.go (hunk 1 above)
Comment (adamreese) [reply cmd/helm/installer/install.go]: That would be a big breaking change for helm.  How long will extensions.Deployments be supported?

Comment (bacongobbler) [reply cmd/helm/installer/install.go]: I'm not sure how long it'll be supported. We should ask at the next SIG-Apps meeting.

Just to clarify, when you mean by breaking change, you mean that `helm init --upgrade` will break by moving to apps/v1, right?

Comment (bacongobbler) [reply cmd/helm/installer/install.go]: From #sig-apps:

> [minimum] 2 releases or 6 months after announced deprecation. See https://kubernetes.io/docs/reference/deprecation-policy/

**Comment by adamreese** [comment]
1.9 is now released

**Review by bacongobbler — APPROVED** [review]

**Comment by bacongobbler** [comment]
Someone else wanna take a look at this over :coffee:?

**Comment by bacongobbler** [comment]
Looks like we got a failing unit test as well in CI. Would you mind rebasing and taking a look into it? Not sure why the linter would fail unless something changed upstream (which is entirely in the realm of plausibility).

**Comment by adamreese** [comment]
@bacongobbler not sure why that extra failed test is on there.  It is in fact passing https://circleci.com/gh/kubernetes/helm/4840?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by bacongobbler** [comment]
Indeed it does seem to be passing. Just as an added measure I checked https://circleci.com/gh/kubernetes/helm/tree/pull%2F3234 and it shows the latest build for this PR as passing, so we can ignore the flake.

File: cmd/helm/installer/install.go (hunk 1 above)
Comment (thomastaylor312) [reply cmd/helm/installer/install.go]: IMO I think we should move to `apps/v1`. Anyone who has upgraded their Kubernetes cluster to 1.9 should know about that switch and we can make it clear in the release notes that for this upgrade they'll need to delete the old Tiller

File: pkg/kube/wait.go (hunk 6 above)
Comment (thomastaylor312) [reply pkg/kube/wait.go]: I agree with @bacongobbler, we'll need to keep these ones and add all of the `apps/v1` types here so as to not break `--wait` functionality

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Most of this is looking great! There is one more change (commented on the existing thread from @bacongobbler) that is needed for sure to make sure Helm supports the new API versions

**Comment by bacongobbler** [comment]
@adamreese would you mind also bumping https://github.com/kubernetes/helm/blob/170cff83c26f2df5ecdf471cb8def846cf69d76d/pkg/chartutil/capabilities.go#L33 to 9?

**Review by mattfarina — COMMENTED** [review]
I ran multiple manual tests and the things appear to be working.

File: cmd/helm/installer/install.go (hunk 1 above)
Comment (technosophos) [reply cmd/helm/installer/install.go]: Yeah, we need to wait, especially as the adoption cycle for new Kubernetes versions is slowing down. (We still see a lot of 1.6 and even 1.5 in prod)

Comment (adamreese) [reply cmd/helm/installer/install.go]: I agree.  I will leave it as it.

**Comment by adamreese** [comment]
@thomastaylor312 updated PTAL

**Review by thomastaylor312 — APPROVED** [review]

**Pull request merged**